### PR TITLE
fixes #8577, #8580, other bug fixes

### LINKED ISSUE #8577: `mapIt` still can't be used with `openArray` even after #8543, #8567: fails during bootstrap
@LemonBoy I was trying to revive my old PR https://github.com/nim-lang/Nim/pull/8272 now that #8543 was merged (Make mapIt work on openArray's #8543) however that PR fails: it was actually kinda confusing since it all works locally (ie when using an already bootstrapped compiler), but the problem is in bootstrap phase:

to reproduce:
```
git clone https://github.com/nim-lang/Nim
cd Nim
git fetch origin pull/8272/head:pr_quoteShellCommand
git checkout pr_quoteShellCommand
sh build_all.sh # (the new install script I added a few commits ago; irrelevant to this bug though)
```

```
/private/tmp/d36/Nim/lib/pure/ospaths.nim(635, 18) template/generic instantiation from here
/private/tmp/d36/Nim/lib/pure/collections/sequtils.nim(680, 14) Error: invalid type: 'openarray[string]' for let
```

for posterity: this is what I was adding and that was failing during bootstrap
```nim
proc quoteShellCommand*(args: openArray[string]): string =
    ## Concatenates and quotes shell arguments `args`
    runnableExamples:
      when defined(posix):
        assert quoteShellCommand(["aaa", "", "c d"]) == "aaa '' 'c d'"
      when defined(windows):
        assert quoteShellCommand(["aaa", "", "c d"]) == "aaa \"\" \"c d\""
    # can't use `map` pending https://github.com/nim-lang/Nim/issues/8303
    result = args.mapIt(quoteShell(it)).join(" ")
```

## links
* #8543
* #8567

### LINKED ISSUE #8580: `let t=s.mapIt(it)` gives `Error: redefinition of 't'`
```nim
#[
BUG:
/Users/timothee/git_clone/nim/timn/bugs/stdlib/t01_mapIt_redefinition_of_t.nim(6, 1) template/generic instantiation from here
/Users/timothee/git_clone/nim/timn/bugs/stdlib/t01_mapIt_redefinition_of_t.nim(9, 9) Error: redefinition of 't'
]#

import sequtils

when defined(case1):
  proc test() =
    # BUG
    let t = [1,2].mapIt(it)

test()
```

@judofyr LemonBoy is there a better fix than changing `t` to something rare in implementation of `mapIt`?

### PR DESCRIPTION
**Pull request by timotheecour**
/c @Araq @LemonBoy more fixes coming after this PR, eg: https://github.com/nim-lang/Nim/pull/8586 (+ 2 others) which depend on this to be merged first ; so can we merge this sooner rather than later?

* fixes #8577
* builds on and improves https://github.com/nim-lang/Nim/pull/8578 (from @LemonBoy) to also fix these:
* fixes #8580
* fixes https://github.com/nim-lang/Nim/pull/8578#issuecomment-411614711 
* avoids duplicate initialization of `result` in `mapIt`
* adds more test cases

### CODE REVIEW

File: lib/pure/collections/sequtils.nim [hunk 1]
```diff
@@ -675,16 +668,16 @@ template mapIt*(s, op: untyped): untyped =
     block:
       var it{.inject.}: type(items(s));
       op))
-  var result: seq[outType]
   when compiles(s.len):
-    evalOnce(t, s)
+    # using t_D20180808T211500 to avoid https://github.com/nim-lang/Nim/issues/8580
+    evalOnce(t_D20180808T211500, s, type(s) isnot openArray)
```
Comment (LemonBoy) [inline lib/pure/collections/sequtils.nim]: This is nasty. Using a `block:` is a somewhat cleaner workaround.

Comment (LemonBoy) [reply lib/pure/collections/sequtils.nim]: Instead of checking whether `s` is an `openArray` I'd just add another `compiles((let _ = s))` to the `when` above.

Comment (Araq) [reply lib/pure/collections/sequtils.nim]: This usage of ``compiles`` seems to be a hack...  :-/

Comment (LemonBoy) [reply lib/pure/collections/sequtils.nim]: Yep, but I can't think of a better way to shunt the bare `openArray` literals into the other code path

Comment (timotheecour) [reply lib/pure/collections/sequtils.nim]: > This is nasty. Using a block: is a somewhat cleaner workaround.

used a block; much cleaner, thanks!

> compiles((let _ = s))

had thought of that but will likely lead to other bugs and hide other issues

> Yep, but I can't think of a better way to shunt the bare openArray literals into the other code path

not sure what you're suggesting besides `compiles` ; what could go wrong with `type(s) isnot openArray` ?

**Comment by timotheecour** [comment]
> One more failing test case: `[].mapIt(it)`

illegal (just as `let a =@[]` is), however added test case for (typed) empty seqs.

PTAL (all comments addressed)

File: lib/pure/collections/sequtils.nim (hunk 1 above)
Comment (timotheecour) [reply lib/pure/collections/sequtils.nim]: as discussed here https://gitter.im/nim-lang/Nim?at=5b6c0cf3637ee66082c775e8, I changed to `compiles((let _ = s))` (see comments in code)

**Comment by timotheecour** [comment]
PTAL all comments addressed (after fix to `compiles((let _ = s))`

File: lib/pure/collections/sequtils.nim [hunk 2]
```diff
@@ -1067,9 +1068,30 @@ when isMainModule:
...
 
   block: # mapIt with openArray
-    when declared(macros.symKind):
-      proc foo(x: openArray[int]): seq[int] = x.mapIt(it + 1)
-      doAssert foo([1,2,3]) == @[2,3,4]
+    proc foo(x: openArray[int]): seq[int] = x.mapIt(it + 1)
+    doAssert foo([1,2,3]) == @[2,3,4]
+
+  block: # mapIt with direct openArray
+    # see https://github.com/_render_node/MDExOlB1bGxSZXF1ZXN0MjA3MTQ3MTIw/pull_requests/timeline#issuecomment-411614711
```
Comment (kaushalmodi) [inline lib/pure/collections/sequtils.nim]: What is the `github.com/_render_node/..` link?

File: lib/pure/collections/sequtils.nim [hunk 3]
```diff
@@ -675,19 +668,27 @@ template mapIt*(s, op: untyped): untyped =
...
-      result[i] = op
-      i += 1
+    block: # using a block avoids https://github.com/nim-lang/Nim/issues/8580
+
+      # avoid double substitution of `s`.
+      # NOTE: if we used `type(s) isnot openArray` instead, the openarray
+      # literal would be substituted twice
+      # BUG: `evalOnce(t, s, false)` would lead to C compile errors
+      # (`error: use of undeclared identifier`) instead of Nim compile errors
+      evalOnce(t, s, compiles((let _ = s)))
```
Comment (dom96) [inline lib/pure/collections/sequtils.nim]: ``t``, ``s``

I know that you didn't introduce these names but I really despise single character variable names. In reviewing this PR I had to go back and forth between various templates to work out what `t` is meant to be here.

Perhaps they could be changed?

Comment (dom96) [reply lib/pure/collections/sequtils.nim]: Side note: this ``evalOnce`` macro feels like a total hack to me.

File: lib/pure/collections/sequtils.nim (hunk 2 above)
Comment (dom96) [reply lib/pure/collections/sequtils.nim]: Yeah... this URL doesn't load.

File: lib/pure/collections/sequtils.nim [hunk 4]
```diff
@@ -1067,9 +1068,30 @@ when isMainModule:
...
   block: # mapIt with openArray
-    when declared(macros.symKind):
-      proc foo(x: openArray[int]): seq[int] = x.mapIt(it + 1)
-      doAssert foo([1,2,3]) == @[2,3,4]
+    proc foo(x: openArray[int]): seq[int] = x.mapIt(it + 1)
+    doAssert foo([1,2,3]) == @[2,3,4]
+
+  block: # mapIt with direct openArray
+    # see https://github.com/_render_node/MDExOlB1bGxSZXF1ZXN0MjA3MTQ3MTIw/pull_requests/timeline#issuecomment-411614711
+    doAssert openArray[int]([1,2]).mapIt(it) == @[1,2]
```
Comment (dom96) [inline lib/pure/collections/sequtils.nim]: A better test would be to create a `proc` that takes an `openarray`. Pass it a list and call mapIt on the argument.

Frankly, I'm surprised this works. ``openArray`` is a pseudo-type AFAIK.

File: lib/pure/collections/sequtils.nim [hunk 5]
```diff
@@ -1067,9 +1068,30 @@ when isMainModule:
...
+
+  block: # mapIt test, see https://github.com/nim-lang/Nim/pull/8584#pullrequestreview-144723468
+    # NOTE: `[].mapIt(it)` is illegal, just as `let a = @[]` is
```
Comment (dom96) [inline lib/pure/collections/sequtils.nim]: What about

```nim
mapIt([], it)

mapIt(@[], it)
```

?

File: lib/pure/collections/sequtils.nim (hunk 4 above)
Comment (timotheecour) [reply lib/pure/collections/sequtils.nim]: i already have a test for that, see "mapIt with openArray"

File: lib/pure/collections/sequtils.nim (hunk 2 above)
Comment (timotheecour) [reply lib/pure/collections/sequtils.nim]: removed it

File: lib/pure/collections/sequtils.nim (hunk 3 above)
Comment (timotheecour) [reply lib/pure/collections/sequtils.nim]: almost procs in this module use single char arg names so I don't wanna special case mapIt, feel free to change that in a subsequent PR.
change s to s2 to make it obvious it relates to s; also improved var names of evalOnce and added a bunch of documentation.

> Side note: this evalOnce macro feels like a total hack to me.

feel free to improve that in a subsequent PR. It's there for a reason (avoid double evaluation bugs, handle openArray and other edge cases) and neither @LemonBoy nor I could come up with something better

File: lib/pure/collections/sequtils.nim (hunk 5 above)
Comment (timotheecour) [reply lib/pure/collections/sequtils.nim]: added a test `doAssert: not compiles(mapIt(@[], it))` ; it should indeed not compile as explained in comments (lacks type information)

**Comment by timotheecour** [comment]
PTAL. all comments addressed.

File: lib/pure/collections/sequtils.nim [hunk 6]
```diff
@@ -675,19 +674,24 @@ template mapIt*(s, op: untyped): untyped =
...
-    var i = 0
-    result = newSeq[outType](t.len)
-    for it {.inject.} in t:
-      result[i] = op
-      i += 1
+    block: # using a block avoids https://github.com/nim-lang/Nim/issues/8580
+
+      # BUG: `evalOnceAs(s2, s, false)` would lead to C compile errors
+      # (`error: use of undeclared identifier`) instead of Nim compile errors
+      evalOnceAs(s2, s, compiles((let _ = s)))
```
[the remark this answers is not in the document]
Comment (timotheecour) [reply lib/pure/collections/sequtils.nim]: I know about this limitation (it's documented in the PR, see below) but I disagree with your proposed change:

* with your suggestion, it'll affect valid (non edge case) code paths, resulting in un-optimized code for these.

please try the PR with your suggestion and see what happens by inserting these echo statements:

```nim
  when compiles(s.len) and compiles((let _ = s)):
    block:
      evalOnceAs(s2, s, true)
      echo "optimized"
      var i = 0
      var result = newSeq[outType](s2.len)
      for it {.inject.} in s2:
        result[i] = op
        i += 1
      result
  else:
    echo "fallback"
    var result: seq[outType] = @[]
    for it {.inject.} in s:
      result.add(op)
    result
```

will print:
```
optimized
fallback
fallback
fallback
optimized
optimized
optimized
optimized
fallback
```

whereas with current PR it prints:
```
optimized
optimized
optimized
optimized
optimized
optimized
optimized
optimized
fallback
```

ie it does the correct thing (optimized code + no duplicate evaluation) in all cases except the in the very edge case of an `openArray` constructor passed to a proc (template would be ok).

copying verbatim from PR:
```nim
  block: # mapIt with direct openArray
    proc foo1(x: openArray[int]): seq[int] = x.mapIt(it * 10)
    counter = 0
    doAssert foo1(openArray[int]([identity(1),identity(2)])) == @[10,20]
    doAssert counter == 2

    template foo2(x: openArray[int]): seq[int] = x.mapIt(it * 10)
    counter = 0
    doAssert foo2(openArray[int]([identity(1),identity(2)])) == @[10,20]
    # TODO: this fails; not sure how to fix this case
    # doAssert counter == 2
```

[the remark this answers is not in the document]
Comment (LemonBoy) [reply lib/pure/collections/sequtils.nim]: Oh I see, that's fine by me then.

[the remark this answers is not in the document]
Comment (timotheecour) [reply lib/pure/collections/sequtils.nim]: cool. PR still marked as @LemonBoy requested changes though...

**Comment by timotheecour** [comment]
PTAL: can we please get this merged ASAP? it fixes pre-existing bugs

**Comment by Araq** [comment]
> it fixes pre-existing bugs

And we're sure it doesn't introduce new bugs, right?

Oh well, will clean this up later.

**Comment by timotheecour** [comment]
thank you. 

> And we're sure it doesn't introduce new bugs, right?

obviously no such guarantees, but it came with increased test coverage
